File: TypeScript/ExampleTests/common/PostString.ts

import * as tokens from "./tokens.js";

export class PostString {
  static readonly DNC_STRING = "?DNC?";

  userId: string;
  id: string;
  title: string;
  body: string;

  constructor(userId: string = "", id: string = "", title: string = "", body: string = "") {
    this.userId = userId;
    this.id = id;
    this.title = title;
    this.body = body;
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text: string): PostString {
    const parts = tokens.require_(text, 4, "Post");
    return new PostString(
      parts[0],
      parts[1],
      parts[2],
      parts[3]
    );
  }

  static fromList(values: Iterable<string>): PostString {
    const v = Array.from(values);
    const r = new PostString();
    r.userId = v[0] ?? "";
    r.id = v[1] ?? "";
    r.title = v[2] ?? "";
    r.body = v[3] ?? "";
    return r;
  }

  toString(): string {
    return tokens.token(this.userId) + " " + tokens.token(this.id) + " " + tokens.token(this.title) + " " + tokens.token(this.body);
  }

  equals(other: PostString): boolean {
    return (this.userId === PostString.DNC_STRING || other.userId === PostString.DNC_STRING || this.userId === other.userId)
      && (this.id === PostString.DNC_STRING || other.id === PostString.DNC_STRING || this.id === other.id)
      && (this.title === PostString.DNC_STRING || other.title === PostString.DNC_STRING || this.title === other.title)
      && (this.body === PostString.DNC_STRING || other.body === PostString.DNC_STRING || this.body === other.body);
  }

  static equalLists(a: PostString[], b: PostString[]): boolean {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]!));
  }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/PostString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class PostString
    {
        public string userId;
        public string id;
        public string title;
        public string body;

        public PostString(string userId, string id, string title, string body)
        {
            this.userId = userId;
            this.id = id;
            this.title = title;
            this.body = body;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static PostString FromText(string text)
        {
            var parts = Tokens.Require(text, 4, "Post");
            return new PostString(parts[0], parts[1], parts[2], parts[3]);
        }

        public PostString(string text)
        {
            var parsed = FromText(text);
            this.userId = parsed.userId;
            this.id = parsed.id;
            this.title = parsed.title;
            this.body = parsed.body;
        }

        public PostTyped ToPostTyped()
        {
            return new PostTyped(
                int.Parse(this.userId),
                int.Parse(this.id),
                this.title,
                this.body
            );
        }

        public override string ToString()
        {
            return Tokens.Token(userId) + " " + Tokens.Token(id) + " " + Tokens.Token(title) + " " + Tokens.Token(body);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not PostString other) return false;
            return (DNCString == this.userId || DNCString == other.userId || object.Equals(this.userId, other.userId))
                && (DNCString == this.id || DNCString == other.id || object.Equals(this.id, other.id))
                && (DNCString == this.title || DNCString == other.title || object.Equals(this.title, other.title))
                && (DNCString == this.body || DNCString == other.body || object.Equals(this.body, other.body));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.userId);
            h.Add(this.id);
            h.Add(this.title);
            h.Add(this.body);
            return h.ToHashCode();
        }
    }
}
